examples/sat/schoening.py

```diff
@@ -1,86 +1,166 @@
 """This module provides the Schöning example for solving SAT instances."""
 
-import itertools
 from typing import Optional
 
+import attrs
 import numpy as np
 from sat import Assignment, SatInstance
 
-from qubrabench.algorithms.search import search
-from qubrabench.benchmark import oracle
+from qubrabench.algorithms.search import (
+    SamplingDomain,
+    search_by_sampling_with_replacement,
+)
+from qubrabench.benchmark import named_oracle, oracle
 
-__all__ = ["schoening_solve"]
+__all__ = ["schoening_solve", "schoening_bruteforce_steps"]
+
+
+@oracle
+def schoening_with_randomness(
+    inst: SatInstance, initial_assignment: np.ndarray, walk_steps: np.ndarray
+) -> Optional[Assignment]:
+    """
+    Run Schoening's algorithm with fixed random choices.
+
+    Args:
+        initial_assignment: starting assignment (bit string)
+        walk_steps: choices to take at each step of the random walk
+        inst: The 3-SAT Instance for which to find a satisfying assignment.
+
+    Returns:
+        Satisfying assignment if found, None otherwise.
+    """
+    assignment = np.copy(initial_assignment)
+
+    # FIXME: get_stats(inst.evaluate) won't work, as underlying matrix can't be hashed.
+    eval = (named_oracle("inst.evaluate"))(lambda k: inst.evaluate(k))
+
+    if eval(assignment):
+        return assignment
+    # iterate over the steps of the random walk
+    for step in walk_steps:
+        # choose an unsatisfied clause
+        unsat_clauses = (inst.clauses @ assignment.T) == -inst.k
+        unsat_clause = inst.clauses[unsat_clauses][0]
+
+        # select a variable that appears in unsatisfied clause, and flip it
+        vars_ = np.argwhere(unsat_clause != 0)
+        var = vars_[step]
+        assignment[var] *= -1
+
+        if eval(assignment):
+            return assignment
+
+    return None
+
+
+@attrs.define
+class SchoeningDomain(SamplingDomain[tuple[np.ndarray, np.ndarray]]):
+    """The class SchoeningDomain implements the methods declared for a SamplingDomain. We use this class for search spaces that cannot be stored in memory."""
+
+    n: int
+    n_assignment: int
+    n_walk_steps: int
+
+    @staticmethod
+    def generate_random_assignment(n, rng: np.random.Generator):
+        return rng.integers(2, size=n) * 2 - 1
+
+    @staticmethod
+    def generate_walk_steps(m, rng: np.random.Generator):
+        return rng.integers(3, size=m)
+
+    def get_size(self) -> int:
+        return 2**self.n_assignment * 3**self.n_walk_steps
+
+    def get_probability_of_sampling_solution(self, key) -> float:
+        return 0.75**self.n
```

Comment:
add todo comment here that this is only soft-O, and is missing poly(n) and constant factors.